packages/codemods/src/v5/keep-previous-data/keep-previous-data.js

```diff
@@ -0,0 +1,208 @@
+
+    try {
+      const firstArgument = node.arguments[0] ?? null
+
+      if (firstArgument && utils.isObjectExpression(firstArgument)) {
+        const isPlaceholderDataPropertyPresent =
+          hasPlaceholderDataProperty(firstArgument)
+
+        if (hasPlaceholderDataProperty(firstArgument)) {
+          throw new AlreadyHasPlaceholderDataProperty(node, filePath)
+        }
+
+        const keepPreviousDataProperty =
+          getKeepPreviousDataProperty(firstArgument)
+
+        const keepPreviousDataPropertyHasTrueValue =
+          isObjectPropertyHasTrueBooleanLiteralValue(keepPreviousDataProperty)
+
+        if (!keepPreviousDataPropertyHasTrueValue) {
+          utils.warn(
+            `The usage in file "${filePath}" at line ${start}:${end} already contains a "keepPreviousData" property but its value is not "true". Please migrate this usage manually.`,
+          )
+
+          return node
+        }
+
+        if (keepPreviousDataPropertyHasTrueValue) {
+          // Removing the `keepPreviousData` property from the object.
+          const mutableObjectExpressionProperties =
+            filterKeepPreviousDataProperty(firstArgument)
+
+          if (!isPlaceholderDataPropertyPresent) {
+            isKeepPreviousDataInUse = true
+
+            // When the `placeholderData` property is not present, the `placeholderData: keepPreviousData` property will be added.
+            mutableObjectExpressionProperties.push(
+              createPlaceholderDataObjectProperty(),
+            )
+          }
+
+          functionArguments.push(
+            jscodeshift.objectExpression(mutableObjectExpressionProperties),
+          )
+
+          return jscodeshift.callExpression(
+            node.original.callee,
+            functionArguments,
+          )
+        }
+      }
+
+      utils.warn(
+        `The usage in file "${filePath}" at line ${start}:${end} could not be transformed, because the first parameter is not an object expression. Please migrate this usage manually.`,
+      )
+
+      return node
+
+      return node
+    } catch (error) {
+      utils.warn(
+        error.name === AlreadyHasPlaceholderDataProperty.name
+          ? error.message
+          : `An unknown error occurred while processing the "${filePath}" file. Please review this file, because the codemod couldn't be applied.`,
+      )
+
+      return node
+    }
+  }
+
+  createUseQueryLikeTransformer({ jscodeshift, utils, root }).execute(
+    config.hooks,
+    replacer,
+  )
+
+  return { isKeepPreviousDataInUse }
+}
+
+module.exports = (file, api) => {
+  const jscodeshift = api.jscodeshift
+  const root = jscodeshift(file.source)
+  const utils = createUtilsObject({ root, jscodeshift })
+  const filePath = file.path
+
+  const dependencies = { jscodeshift, utils, root, filePath }
+
+  const { isKeepPreviousDataInUse } = transformUsages({
+    ...dependencies,
+    config: {
+      hooks: ['useInfiniteQuery', 'useQueries', 'useQuery'],
```

Comment:
@TkDodo other hooks might be affected?

Reply:
oops I missed this comment. We also have the global defaults like:

```
new QueryClient({
  defaultOptions: {
    queries: {
      keepPreviousData: true
    }
  }
})
```

and: 

```
queryClient.setQueryDefaults(['key'], {
  keepPreviousData: true
})
```

Reply:
Don't worry, I will open another PR.

Reply:
Covered by:
- https://github.com/TanStack/query/pull/6283